Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -245,6 +245,65 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
         parseWebAccessibleResourcesVersion2();
 }
 
+URL WebExtension::resourceFileURLForPath(const String& originalPath)
+{
+    ASSERT(originalPath);
+
+    String path = originalPath;
+
+    if (path.startsWith('/'))
+        path = path.substring(1);
+
+    if (!path.length() || m_resourceBaseURL.isNull())
+        return { };
+
+    auto resourceURL = URL(m_resourceBaseURL, path);
+
+    // Don't allow escaping the base URL with "../".
+    if (!resourceURL.string().startsWith(m_resourceBaseURL.string())) {
+        RELEASE_LOG_ERROR(Extensions, "Resource URL path escape attempt: %s", resourceURL.string().utf8().data());
+        return { };
+    }
+
+    return resourceURL;
+}
+
+String WebExtension::resourceStringForPath(const String& unsafePath, RefPtr<API::Error>& outError, CacheResult cacheResult, SuppressNotFoundErrors suppressErrors)
+{
+    ASSERT(unsafePath);
+
+    String path = unsafePath;
+
+    // Remove leading slash to normalize the path for lookup/storage in the cache dictionary.
+    if (path.startsWith('/'))
+        path = path.substring(1);
+
+    if (auto cachedData = m_resources.getOptional(path)) {
+        if (auto cachedString = String::fromUTF8(*cachedData); !cachedString.isEmpty())
+            return cachedString;
+    }
+
+    bool isServiceWorker = backgroundContentIsServiceWorker();
+    if (!isServiceWorker && (path == generatedBackgroundPageFilename))
+        return generatedBackgroundContent();
+
+    if (isServiceWorker && (path == generatedBackgroundServiceWorkerFilename))
+        return generatedBackgroundContent();
+
+    auto data = resourceDataForPath(path, outError, CacheResult::No, suppressErrors);
+    auto resourceString = emptyString();
+    if (!data.empty())
+        resourceString = String::fromUTF8(data);
```

Comment:
We can't assume UTF8 here. The previous code guessed the string encoding. WTF should have something similar.

Reply:
Unless I'm missing something, it doesn't look like there's a way to guess with WTF::String. There's a function that falls back to Latin-1 though, which could be better here, but I don't know if another encoding is required.